src/btree/row_srch.c

```diff
@@ -116,6 +127,40 @@ __wt_search_insert(
             last_ins = ins;
             key.data = WT_INSERT_KEY(ins);
             key.size = WT_INSERT_KEY_SIZE(ins);
+            /*
+             * We have an optimization to reduce the number of bytes we need to compare during the
+             * search if we know a prefix of the search key matches the keys we have already
+             * compared on the upper stacks. This works because we know the keys become denser down
+             * the stack.
+             *
+             * However, things become tricky if we have another key inserted concurrently to next of
+             * the search key. If we have seen this key inserted upper the stack, we require the
+             * insertion must also be visible to us down the stack. Otherwise, we may wrongly skip
+             * the comparison of a prefix and land on the wrong spot. Here's an example:
+             *
+             * Suppose we have a skip list:
+             *
+             * L1: AA -> BD
+             *
+             * L0: AA -> BD
+             *
+             * and we want to search AB and a key AC is inserted concurrently. If we see the
+             * following skip list in the search:
+             *
+             * L1: AA -> AC -> BD
+             *
+             * L0: AA -> BD
+             *
+             * Since we have compared with AA and AC on level 1 before dropping down to level 0, we
+             * decide we can skip comparing the first byte of the key. However, since we don't see
+             * AC on level 0, we compare with BD and wrongly skip the comparison with prefix B.
+             *
+             * On architectures with strong memory ordering, the requirement is satisfied by
```

Comment:
This might need a little bit more elaboration or a correction. ARM's ```CASAL``` is also a full barrier comparable to x86_64's ```LOCK CMPXCHG```. What is different is that reads are not re-ordered against other reads on x86, whereas on weaker memory models they could be. In both cases, a full barrier would clear the cache line and synchronize it to the latest values. But that doesn't prevent us on ARM from having already read L0 before the CAS operation and then having the "stale" information.

Reply:
> a full barrier would clear the cache line and synchronize it to the latest values

That isn't true at all. A) all modern CPUs are fully cache coherent using something like [MOESI](https://en.wikipedia.org/wiki/MOESI_protocol), so if a cache read succeeds, you are guaranteed to get its "latest value" (I'm ignoring things like mapped mapped I/O and non-CPU devices writing to memory, since that is more a concern for kernels than us). Barriers don't have to do any sort of cache-line clearing, since that is done automatically as part of normal reads and writes. The issue is that superscalar cores (which is basically of them at this point) may have multiple reads or writes in flight at once and without fencing, they may complete in an order that doesn't match the program order. 

Note that even x86 isn't fully ordered and allows loads to complete prior to (plain) stores that are earlier in program order (Run [this](https://gist.github.com/RedBeard0531/b8b0a7b41e7bdaddbbddb9d226df2452#file-sc-cpp-L88-L101) on x86 if you want proof). This is due to store buffers which allow a write instruction to "complete" before it makes it to cache and becomes globally visible.

But none of this really matters. We shouldn't be talking about the hardware implementation, since that isn't part of the platform guarantees. We should instead be able to justify correctness by pointing to what we do to ensure that the memory model guarantees us correct ordering  where we need it. In this case, we need to read the L(n+1) pointer prior to reading the L(n) pointer. In the old code, we had a bug because we did not ensure that the reads were ordered correctly and the fix is to correctly order the reads. I think the only thing that needs explaining is _why_ we care about the order. We should accept that `WT_ORDERED_READ` correctly enforces ordering and take to heart that we must always do _something_ to enforce order for both reads and writes whenever we care about it. We shouldn't need to talk about strong and weak memory models and specific instructions in general code because it shouldn't care about that. It just needs to use our preferred mechanisms to ensure that ordering.

Reply:
I removed the word about detailed hardware implementation. But I think we still need to mention about the hardware architecture because wiretiger is not supporting all the hardware platforms and we need to be aware of there differences.

Reply:
@RedBeard0531 , thanks for giving more insight into cache coherence on modern CPUs. I agree that talking about hardware implementations is a distraction we should stay away from. Until we follow the memory access order required for correctness, we don't need to be any wiser about how each platform manages that.